File: mobile/lib/models/carrito.dart
import '../core/api_config.dart';

/// Espejo de `CarritoOut`/`DetalleCarritoOut` (backend, CU11). El carrito
/// ahora vive en el backend ligado al cliente -- igual que en la web desde
/// CU11, ya no es local. Pydantic serializa Decimal como texto (ej.
/// "130.00"), por eso se parsea con `num.parse` en vez de asumir `num`.
class DetalleCarrito {
  const DetalleCarrito({
    required this.id,
    required this.productoId,
    required this.productoNombre,
    this.productoImagenUrl,
    required this.tallaId,
    required this.tallaCodigo,
    required this.colorId,
    required this.colorNombre,
    required this.cantidad,
    required this.precioUnitario,
    required this.subtotal,
  });

  final int id;
  final int productoId;
  final String productoNombre;
  final String? productoImagenUrl;
  final int tallaId;
  final String tallaCodigo;
  final int colorId;
  final String colorNombre;
  final int cantidad;
  final double precioUnitario;
  final double subtotal;

  factory DetalleCarrito.fromJson(Map<String, dynamic> json) {
    return DetalleCarrito(
      id: json['id'] as int,
      productoId: json['producto_id'] as int,
      productoNombre: json['producto_nombre'] as String,
      productoImagenUrl: (json['producto_imagen_url'] as String?) != null
          ? resolveImageUrl(json['producto_imagen_url'] as String)
          : null,
      tallaId: json['talla_id'] as int,
      tallaCodigo: json['talla_codigo'] as String,
      colorId: json['color_id'] as int,
      colorNombre: json['color_nombre'] as String,
      cantidad: json['cantidad'] as int,
      precioUnitario: num.parse(json['precio_unitario'].toString()).toDouble(),
      subtotal: num.parse(json['subtotal'].toString()).toDouble(),
    );
  }
}

class Carrito {
  const Carrito({required this.id, required this.estado, required this.detalles, required this.total});

  final int id;
  final String estado;
  final List<DetalleCarrito> detalles;
  final double total;

  int get totalItems => detalles.fold(0, (sum, d) => sum + d.cantidad);

  factory Carrito.fromJson(Map<String, dynamic> json) {
    return Carrito(
      id: json['id'] as int,
      estado: json['estado'] as String,
      detalles: (json['detalles'] as List<dynamic>)
          .map((e) => DetalleCarrito.fromJson(e as Map<String, dynamic>))
          .toList(),
      total: num.parse(json['total'].toString()).toDouble(),
    );
  }
}

File: backend/app/schemas/carrito.py
from datetime import datetime, timezone
from decimal import Decimal

from pydantic import BaseModel, Field, field_serializer


class CarritoItemCreate(BaseModel):
    producto_id: int
    talla_id: int
    color_id: int
    sucursal_id: int
    cantidad: int = Field(ge=1)


class CarritoItemUpdate(BaseModel):
    cantidad: int = Field(ge=1)


class DetalleCarritoOut(BaseModel):
    id: int
    producto_id: int
    producto_nombre: str
    producto_imagen_url: str | None
    talla_id: int
    talla_codigo: str
    color_id: int
    color_nombre: str
    sucursal_id: int
    sucursal_nombre: str
    cantidad: int
    precio_unitario: Decimal
    subtotal: Decimal


class CarritoOut(BaseModel):
    id: int
    estado: str
    detalles: list[DetalleCarritoOut]
    total: Decimal


class CarritoAdminOut(CarritoOut):
    cliente_id: int
    cliente_nombre: str
    cliente_email: str
    fecha_creacion: datetime
    fecha_actualizacion: datetime
    cantidad_items: int

    @field_serializer("fecha_creacion", "fecha_actualizacion")
    def _serialize_fecha(self, value: datetime) -> str:
        if value.tzinfo is None:
            value = value.replace(tzinfo=timezone.utc)
        return value.isoformat()
